server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java

```diff
@@ -90,149 +84,19 @@ public void process(HitContext hit) throws IOException {
                         // docValues fields will still be document fields, and put under "fields" section of a hit.
                         hit.hit().setDocumentField(f.field, hitField);
                     }
-                    f.setValues(hit.docId(), hitField);
+                    hitField.getValues().addAll(f.fetcher.fetchValues(hit.sourceLookup()));
                 }
             }
         };
     }
 
-    private abstract static class DocValueField {
```

Comment:
All of the "brains" of this stuff has moved into the `LeafFieldData` implementations.